### ISSUE TITLE
test set mAP higher than val mAP

### ISSUE THREAD

**Issue by EslamHaourn** [ultralytics/yolov5#8415]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello everyone,
Thanks for this great repository.
I have a problem in training on KITTI dataset.

When I use one portion of data for val and test set and the other portion for training(splitting the data as following):
train: ../kitti(augment3classes_crop)/images/train/ # 5751 images
val: ../kitti(val_augment3classes_crop)/images/train/ # 5751 images
test: ../kitti(val_augment3classes_crop)/images/train/ # 1730
I got the same mAP for training and testing.

When I use one portion of data for train and val set and the other for test(splitting the data as following):
train: ../kitti(augment3classes_crop)/images/train/ # 5751 images
val: ../kitti(augment3classes_crop)/images/train/ # 5751 images
test: ../kitti(test_augment3classes_crop)/images/train/ # 1730
I got an overfit 85 mAP for training and 76 mAP for testing. When I use dropout or reduce the momentum value, the two mAPs decreases and the overfit remain. When I increase the data with the BDD dataset the two mAPs decrease more.

When I split the data train, val and test sets as following:
train: ../kitti(augment3classes_crop)/images/train/ # 5751 images
val: ../kitti(val_augment3classes_crop)/images/train/ # 5751 images
test: ../kitti(test_augment3classes_crop)/images/train/ # 1730 
I got the same mAP for train and val sets 76 and I got 77 mAP for the test set.

Another issue some classes with higher training images and instances has lower mAP.

Thanks in advance.

**Comment by glenn-jocher** [comment]
@EslamHaourn 76 and 77 are basically the same value.

**Comment by EslamHaourn** [comment]
> @EslamHaourn 76 and 77 are basically the same value.

Thanks for your reply.
What about some classes that have higher training images and instances has but lower mAP?

**Comment by glenn-jocher** [comment]
@EslamHaourn well out of domain data may not help. Some pointers:

<img width="2029" alt="dataset" src="https://user-images.githubusercontent.com/26833433/176711684-97fcefb9-41ef-4508-a1fb-6ab3677c583a.png">

**Comment by EslamHaourn** [comment]
I'm not sure if I understand the image but the KITTI dataset doesn't have night images for example, however when I add night images from BDD dataset the mAP the overall decreases.
Execuse me, shouldn't test set has lower mAP than val set?

**Comment by glenn-jocher** [comment]
> Execuse me, shouldn't test set has lower mAP than val set?

No

**Comment by EslamHaourn** [comment]
> I'm not sure if I understand the image but the KITTI dataset doesn't have night images for example, however when I add night images from BDD dataset the mAP the overall decreases.
> Execuse me, shouldn't test set has lower mAP than val set?



To what extent might test set mAP allowed to surpass the val set mAP?
I'm not sure if I understand the image but the KITTI dataset doesn't have night images for example, however when I add night images from BDD dataset the mAP the overall decreases.

**Comment by glenn-jocher** [comment]
@EslamHaourn to no extent

**Comment by EslamHaourn** [comment]
Could I know please, why val set appears in training process?
Many thanks.

**Comment by glenn-jocher** [comment]
Training uses val set to validate after every epoch.

**Comment by EslamHaourn** [comment]
> Training uses val set to validate after every epoch.
Validate what? Shouldn't val set not seen in training and val set not seen in testing?

**Comment by glenn-jocher** [comment]
@EslamHaourn train set is used for updating the model, val set for validating during training and for saving best.pt.

Test set is optional and not used during training at all. Sometimes a test set is provided without labels to send results into a private evaluation server in ML competitions etc.

**Comment by EslamHaourn** [comment]
> @EslamHaourn train set is used for updating the model, val set for validating during training and for saving best.pt.

How it can be done at the same time and at the same file. Then what does val.py do with val set differ than train.py. What is the difference between two tasks val and set in val.py? 

Could you help me know what is 'validating during training' called in academia?

**Comment by glenn-jocher** [comment]
See https://en.wikipedia.org/wiki/Training,_validation,_and_test_data_sets
